xcp_d/interfaces/prepostcleaning.py

```diff
@@ -568,63 +568,114 @@ def _run_interface(self, runtime):
+            "The name of the modified file to write out. "
+            "prepared_timeseries.dtseries.nii by default."
+        ),
     )
 
 
-class _CiftiZerosToNaNsOutputSpec(TraitedSpec):
+class _CiftiPrepareForParcellationOutputSpec(TraitedSpec):
     out_file = File(exists=True, mandatory=True, desc="Output CIFTI file.")
 
 
-class CiftiZerosToNaNs(SimpleInterface):
-    """Convert all all-zero vertices' time series to NaNs in a CIFTI file.
+class CiftiPrepareForParcellation(SimpleInterface):
+    """Apply 50% coverage threshold to parcellated data."""
 
-    This is done so that these vertices will be flagged as missing data by wb_command.
-    This interface is only designed to work with dtseries CIFTIs where the first axis is time and
-    the second is space.
-    """
-
-    input_spec = _CiftiZerosToNaNsInputSpec
-    output_spec = _CiftiZerosToNaNsOutputSpec
+    input_spec = _CiftiPrepareForParcellationInputSpec
+    output_spec = _CiftiPrepareForParcellationOutputSpec
 
     def _run_interface(self, runtime):
-        cifti_obj = nb.load(self.inputs.in_file)
-        # load data as memmap
-        data = cifti_obj.get_fdata()
-        # load it in memory
-        data = np.array(data)
-
-        # find all vertices with all zeros or one or more NaNs
-        stdevs = np.std(data, axis=0)
-        # nan > 0 == False, nan <= 0 == False
-        zero_std = ~(stdevs > 0)  # std over time is zero or NaN
-        zero_values = ~(data[0, :] > 0)  # first time point's value is zero or NaN
-        bad_vertex_idx = np.where(np.logical_and(zero_std, zero_values))[0]
-
-        if bad_vertex_idx.size:
+        data_file = self.inputs.data_file
+        atlas_file = self.inputs.atlas_file
+        TR = self.inputs.TR
+        out_file = self.inputs.out_file
+
+        assert data_file.endswith(".dtseries.nii")
+        assert atlas_file.endswith(".dlabel.nii"), atlas_file
+
+        data = read_ndata(data_file)
+        atlas_data = read_ndata(atlas_file)
+        atlas_data = np.squeeze(atlas_data)  # squeeze to 1D
+        assert atlas_data.ndim == 1
+        assert atlas_data.shape[0] == data.shape[0]  # same number of vertices
+
+        bad_vertices_idx = np.where(np.all(np.logical_or(data == 0, np.isnan(data)), axis=1))
+
+        # First, replace all bad vertices' time series with NaNs.
+        # This way, any partially-covered nodes will have NaNs in the bad portions,
+        # so those vertices will be ignored by wb_command -cifti-parcellate.
+        data[bad_vertices_idx, :] = np.nan
+
+        # The problem is that wb_command -cifti-parcellate will raise an error if a whole parcel is
+        # NaNs, so we must replace entirely-bad parcels with zeros instead.
+        n_partially_covered_nodes, n_poorly_covered_nodes, n_uncovered_nodes = 0, 0, 0
+        parcel_ids = np.unique(atlas_data)[1:]
+        n_nodes = parcel_ids.size
+        for i_parcel in parcel_ids:
+            parcel_idx = np.where(atlas_data == i_parcel)[0]
```

Comment:
some more comments in this section would be helpful

Reply:
Done!